lib/imagemounter/index.js

```diff
@@ -0,0 +1,120 @@
+        throw err;
+    }
+    if (fileStat.isDirectory()) {
+        throw new Error(`The provided ${fileType} path is expected to be a file, but a directory was given: ${filePath}`);
+    }
+    return fileStat;
+}
+
+class ImageMounter extends BaseServicePlist {
+
+    /**
+     * Lookup for mounted images.
+     * @param {string} imageType Type of image, `Developer` by default.
+     * @returns {Promise<Buffer[]>} Signature of each mounted image.
+     */
+    async lookup(imageType = 'Developer') {
+        const ret = await this._plistService.sendPlistAndReceive({
+            Command: 'LookupImage',
+            ImageType: imageType
+        });
+        return checkIfError(ret).ImageSignature || [];
+    }
+
+    /**
+     * Check if developer image is mounted.
+     */
+    async isDeveloperImageMounted() {
+        return (await this.lookup()).length > 0;
+    }
+
+    /**
+     * Mount image for device.
+     * @param {string} imageFilePath The file path of image.
+     * @param {string} imageSignatureFilePath The signature file path of given `DeveloperDiskImage.dmg`
+     * @param {string} imageType Type of image, `Developer` by default.
+     */
+    async mount(imageFilePath, imageSignatureFilePath, imageType = 'Developer') {
+        //check file stats
+        const imageFileStat = await assertIsFile(imageFilePath, FILE_TYPE_IMAGE);
+        await assertIsFile(imageSignatureFilePath, FILE_TYPE_SIGNATURE);
+        //read signature
+        const signature = await readFile(imageSignatureFilePath);
+        const mountedImages = await this.lookup(imageType);
+        if (mountedImages.find((mountedSignature) => signature.equals(mountedSignature))) {
+            log.info(`An image with same signature of ${imageSignatureFilePath} is mounted. Doing nothing here`);
+            return;
+        }
+        //notify device
+        const imageSize = imageFileStat.size;
+        const receiveBytesResult = await this._plistService.sendPlistAndReceive({
+            Command: 'ReceiveBytes',
+            ImageSignature: signature,
+            ImageSize: imageSize,
+            ImageType: imageType
+        });
+        if (checkIfError(receiveBytesResult).Status !== 'ReceiveBytesAck') {
+            throw new Error(`Unexpected return from ${MOBILE_IMAGE_MOUNTER_SERVICE_NAME} on sending ReceiveBytes: ${JSON.stringify(receiveBytesResult)}`);
+        }
+        //push image to device
+        const stream = createReadStream(imageFilePath);
+        try {
+            await new B((resolve, reject) => {
+                stream.on('end', resolve);
+                stream.on('error', reject);
+                stream.on('data', async (data) => {
+                    try {
+                        await this._plistService._socketClient.write(data);
+                    } catch (e) {
+                        stream.emit('error', e);
+                    }
+                });
+            });
+        } finally {
+            stream.close();
```

Comment:
I believe the stream gets closed automatically unless autoClose option is set to false explicitly

Reply:
even if an exception is thrown? I didn't know this. If we are sure about that, then this might be redundant

Reply:
I don't have any strong opinion on that. The current approach is also ok